### PR TITLE
fixes #18932 - Introducing Discovery kind of facts

### PR DESCRIPTION
**Pull request by domitea**
Fixing bug in Discovery where all facts from dicovered hosts are recognised as puppet facts. So this PR solved this problem by creating Discovery kind of facts for non-Puppet environments.

The code problem was in `FactImporterRegistry` where is defined only one importer `PuppetFactImporter` and this importer as marker as default. That means that this importer is used every time when there no importer that fits better for given fact. Because Discovery doesn't have any own importer, every fact from Discovery is taken as puppet fact. Thanks to that the `FactName` is type of 'PuppetFactName' and Puppet fact is created and recognized even there no puppet installed at all.

So to avoid this, I created a `DiscoveryFactImporter` with own `DiscoveryFactName`. This will keep facts of discovered hosts as discovery type. However, this PR is targeted to non-puppet environments so I'm not sure if it's work properly with puppet. 

Also, there is missing a piece that could transform a puppet like facts from discovery into discovery like facts. And because it's change of type of `FactName`, the migration will be easy to write. However, only for non-puppet environments. So it will be better to do migration for that, or a task?

### CODE REVIEW

File: app/models/discovery_fact_name.rb
Comment (domitea) [inline app/models/discovery_fact_name.rb:7]: For recognition purposes I used a icon used for Darwin based operating system recognition. I'm not sure if it's good to stay with that or use/create another one.

Comment (ares) [reply app/models/discovery_fact_name.rb:7]: I think that's perfectly ok, we don't have a specific discovery icon

File: app/services/discovery_fact_importer.rb
Comment (ares) [inline app/services/discovery_fact_importer.rb]: Will this work or is this even necessary? The proxy running on FDI is typically not registered in Foreman, but I have no idea how authorization works in this case of facts upload. Perhaps we're skipping it entirely?

Comment (domitea) [reply app/services/discovery_fact_importer.rb]: I just copied a `PuppetFactImporter`, so that's the reason, why this method has been created. I tried that without this method and `DiscoveryFactImporter` working properly. So I deleted this method and updated PR.

**Comment by domitea** [comment]
@ares Tests are not failing, so I think it's ready for merge.

File: app/models/discovery_fact_name.rb
Comment (rabajaj0509) [reply app/models/discovery_fact_name.rb:7]: Although, can we use the `gear` icon, just to keep it consistent with the welcome page?

Comment (rabajaj0509) [reply app/models/discovery_fact_name.rb:7]: Although, can we use the gear icon, just to keep it consistent with the welcome page?

Comment (domitea) [reply app/models/discovery_fact_name.rb:7]: Well, I would like to use a some icon with D in the center (like Puppet one). I'm not sure what kind of icon do you think but I really don't care what kind icon it will be. I only want to have a D in there because it signals that fact came from Discovery.

What do you think? @rabajaj0509 @ares

Comment (lzap) [reply app/models/discovery_fact_name.rb:7]: I am okay with D or Foreman icon. Not sure about gears, that looks weird compared to other facts on a single page.

**Review by lzap — CHANGES_REQUESTED** [review]
It works if you start with a clean FactName table. But once you already have fact names with type "PuppetFactName" and apply this patch, after new host uploads facts I am unable to fetch them from the database for some reason:

http://localhost:3000/fact_values?search=name+%3D+architecture&page=1

This no longer shows any results, there must be some (default) scope that limits it.

**Comment by lzap** [comment]
Please test this scenario, maybe add a test:

* Add a FactName with PuppetFactName type
* Import a host via discovery
* Search facts

We really must not break this.

**Comment by domitea** [comment]
This bad behavior can be archived with the `develop` branch too. So I think that my PR doesn't broke that. Yes, It should be fixed but It's out of scope of this PR.

**Comment by lzap** [comment]
Okay, I would like to do one more change. This is a great opportunity to update a behavior - method `set_build_clean_facts` is overriden to prevent deletion of facts which starts with `discovery_`. Can you modify the behaviour so it actually prevents deleting of all facts stored as Discovery origin.

Also fix the tests and include a case when a discovered fact is named without the prefix.

**Comment by domitea** [comment]
Checking if only one test is failed...

File: app/services/foreman_discovery/host_converter.rb
Comment (timogoebel) [inline app/services/foreman_discovery/host_converter.rb]: This could be refactored to use pure rails notation.
```ruby
FactValue.where.not(id: keep_ids)
```
Even better would be to use subqueries.

Comment (domitea) [reply app/services/foreman_discovery/host_converter.rb]: Thanks for the idea! Select is refactored.

**Comment by domitea** [comment]
@lzap I've updated the tests. However, since this PR we're not rely on the name of fact, the fact can have any name, right?

File: app/services/foreman_discovery/host_converter.rb
Comment (timogoebel) [inline app/services/foreman_discovery/host_converter.rb]: You should be able to omit the pluck here, see

https://github.com/rubocop-hq/rubocop-rails/issues/246

Comment (rabajaj0509) [reply app/services/foreman_discovery/host_converter.rb]: Nit: Since there are so many methods involved, it will be difficult to debug in the future if anything went wrong. Lets break it to getting the `id` in one and then deleting it in the next. What do you think @domitea?

Comment (domitea) [reply app/services/foreman_discovery/host_converter.rb]: Thanks @timogoebel for tip what to omit.

I understand, that huge one liner can be difficult to debug @rabajaj0509. So I split that into two lines.

**Comment by rabajaj0509** [comment]
Tests are still failing @domitea

**Review by rabajaj0509 — APPROVED** [review]
LGTM! :+1: 

Let's wait for @lzap's final approval :)

**Comment by ares** [comment]
@lzap any more comments?

File: test/unit/host_discovered_test.rb
Comment (lzap) [inline test/unit/host_discovered_test.rb]: Please do not delete this test but rewrite it so the behavior is actually tested. What you want to do is to create some facts of a different type manually (create `FactValue` and `FactName` via FactoryBot) and test if these were actually removed.

**Review by lzap — CHANGES_REQUESTED** [review]
Sorry for the delay, the code is okay, however we should not remove a test for the deletion.

File: app/services/foreman_discovery/host_converter.rb
Comment (lzap) [inline app/services/foreman_discovery/host_converter.rb:30]: Apologies I misled you here, actually purpose of discovery_clean_facts is to clean facts starting with discovery_ which are some facts which can be useful during provisioning. Fact type is actually irrelevant, can you put this hunk back as it was? Of course the code must search DiscoveryFactName but also starting with discovery_.

Sorry my bad, I told you to do this change but now I see it is incorrect.

**Comment by domitea** [comment]
I've updated a PR with your comment, @lzap. I hope I merged that functionality well...

File: app/services/foreman_discovery/host_converter.rb
Comment (lzap) [reply app/services/foreman_discovery/host_converter.rb:30]: Nono, what I mean is: delete all facts which are type of DiscoveryFactName AND they start with "discovery_" prefix.

**Review by lzap — CHANGES_REQUESTED** [review]
I am not helping much, am I? :)

**Comment by domitea** [comment]
@lzap I will took the the quote from The Mandalorian: "This is the way" and I only add ", the hard way" :D

I hope that the deleting of facts is finally correct.

File: app/services/foreman_discovery/host_converter.rb
Comment (lzap) [inline app/services/foreman_discovery/host_converter.rb:27]: Have you removed the `pluck` call on purpose? It is way better in this case (faster, less memory used).

File: test/unit/host_discovered_test.rb
Comment (lzap) [inline test/unit/host_discovered_test.rb]: Can you revert this test completely to the original state? If I am not mistaken, no change should be required in this case so it should be passing. This way we will be sure we are not changing any behavior.

**Review by lzap — CHANGES_REQUESTED** [review]
Kindof :-)

**Review by lzap — APPROVED** [review]
It works fine, can you do Upgrade note:

Discovered hosts have a new kind of fact (discovery) so facts will no longer appear on the Monitor - Facts page. To search for discovery hosts, use the search box on the Hosts - Discovered page.

We usually do that in theforeman.org github repository, example: https://github.com/theforeman/theforeman.org/pull/1813

Thanks!

**Pull request merged**
